### PR TITLE
Fuse withdraw sentinel

### PR DESCRIPTION
**Pull request by Joeysantoro**
# PR Type: Feature

### PR Description: 
A sentinel that automatically detects available liquidity in a Fuse pool and withdraws, with a configurable minimum amount left.

### PR Checklist - Feature (Non-Proposal)

- [x] Enhance testing
- [x] All Tests Passing
- [x] Remove Any .only's on Tests

### CODE REVIEW

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        return false;
+    }
+
+    /// @notice return the amount that can be withdrawn from a deposit after leaving min liquidity
+    function getAmountToWithdraw(ERC20CompoundPCVDeposit deposit) public view returns (uint256) {
+        // can't read underlying directly because some of the PCV deposits use an old abi
+        IERC20 underlying = IERC20(withdrawInfos[address(deposit)].underlying);
+        // Reserves of underlying left in the cToken are considered withdrawable liquidity
+        uint256 liquidity = underlying.balanceOf(address(deposit.cToken()));
```
Comment (eswak) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:78]: is that correct ? ctokens should be holding the platform fees & reserves too ?

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    function setWithdrawInfo(address deposit, WithdrawInfo calldata withdrawInfo)
+        public
+        onlyTribeRole(TribeRoles.GUARDIAN)
+    {
+        withdrawInfos[deposit] = withdrawInfo;
+    }
+
+    /// @notice helper method to trustlessly and publicly remove deposits once emptied.
+    /// TODO could be removed to simplify/minimize attack surface and public functions.
+    function removeDepositIfEmpty(ERC20CompoundPCVDeposit deposit) public {
```
Comment (thomas-waite) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol]: Don't have strong feelings on this. Would probably lean on the side of removing/restricting to only the guardian to minimise the surface area

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        address[] memory destinations,
+        address[] memory underlyings,
+        uint96[] memory liquidityToLeaveList
+    ) CoreRef(core) {
+        uint256 len = deposits.length;
+        require(len == destinations.length && len == liquidityToLeaveList.length && len == underlyings.length);
+        for (uint256 i = 0; i < len; ) {
+            fuseDeposits.add(deposits[i]);
+            withdrawInfos[deposits[i]] = WithdrawInfo({
+                destination: destinations[i],
```
Comment (thomas-waite) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:44]: Is it worthwhile checking that all destinations are `safeAddresses`? `pcvGuardian.isSafeAddress()`? Then can guarantee we can recover the transferred out funds and not make a mistake by setting an incorrect address

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            }
+        }
+    }
+
+    /// @notice setter for the Fuse deposit destination and minimum liquidity
+    function setWithdrawInfo(address deposit, WithdrawInfo calldata withdrawInfo)
+        public
+        onlyTribeRole(TribeRoles.GUARDIAN)
+    {
+        withdrawInfos[deposit] = withdrawInfo;
```
Comment (thomas-waite) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:59]: Would probably check the destination is a safe address

**Review by thomas-waite — COMMENTED** [review]
Really like the idea of this overall. I would check that the destinations are safe addresses probably to make sure there's no accidental mistake in where the recovered liquidity is sent

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol (hunk 3 above)
Comment (Joeysantoro) [reply contracts/sentinel/guards/FuseWithdrawalGuard.sol:44]: PCV Guardian can only transfer to a safe address so this is baked in

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol (hunk 4 above)
Comment (Joeysantoro) [reply contracts/sentinel/guards/FuseWithdrawalGuard.sol:59]: this is done by PCV guardian

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol (hunk 2 above)
Comment (Joeysantoro) [reply contracts/sentinel/guards/FuseWithdrawalGuard.sol]: Will remove

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol (hunk 1 above)
Comment (Joeysantoro) [reply contracts/sentinel/guards/FuseWithdrawalGuard.sol:78]: They are withdrawable either way. Platform fees do not have a liquidity preference

File: contracts/test/integration/sentinel/FuseWithdrawalGuard.t.sol [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        uint96[] memory liquidityToLeaveList = new uint96[](len);
+
+        deposits[0] = MainnetAddresses.RARI_POOL_8_FEI_PCV_DEPOSIT;
+        deposits[1] = MainnetAddresses.RARI_POOL_8_LUSD_PCV_DEPOSIT;
+
+        underlyings[0] = MainnetAddresses.FEI;
+        underlyings[1] = MainnetAddresses.LUSD;
+
+        destinations[0] = MainnetAddresses.DAI_PSM;
+        destinations[1] = MainnetAddresses.LUSD_PSM;
```
Comment (Joeysantoro) [inline contracts/test/integration/sentinel/FuseWithdrawalGuard.t.sol:39]: if LUSD psm is deprecated, this needs to change to the native deposit

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        uint96 liquidityToLeave;
+    }
+
+    /// @notice map the destination and minimum liquidity for each pcv deposit
+    mapping(address => WithdrawInfo) public withdrawInfos;
+
+    EnumerableSet.AddressSet private fuseDeposits;
+
+    /// @notice the PCV mover contract exposed to guardian role
+    PCVGuardian public constant pcvGuardian = PCVGuardian(0x02435948F84d7465FB71dE45ABa6098Fc6eC2993);
```
Comment (xklob) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:25]: Would prefer if this address was declared as a constant at the top of the file

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    /// @notice added to prevent dust from bricking the contract
+    uint256 public constant MIN_WITHDRAW = 100e18;
```
Comment (xklob) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:30]: Unclear naming - is this min ctoken withdrawal, or underlying? Might want to update var name or comment.

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    constructor(
+        address core,
+        address[] memory deposits,
+        address[] memory destinations,
+        address[] memory underlyings,
+        uint96[] memory liquidityToLeaveList
+    ) CoreRef(core) {
```
Comment (xklob) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:38]: Why is this coreRef? Guards ideally shouldn't rely on external state, which CoreRef brings in.

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        // max withdraw is the pcv deposit balance
+        uint256 withdrawAmount = deposit.balance();
+        if (withdrawAmount > liquidity) {
+            withdrawAmount = liquidity;
+        }
+        return withdrawAmount > MIN_WITHDRAW ? withdrawAmount : 0;
+    }
+
+    /// @notice return the first element which can be withdrawn from with the appropriate calldata encoding tha max withdraw amount.
+    /// @dev it only returns one element for simplicity. Unlikely multiple will be withdrawable simulteneously after the first pass, and pruning empty entries from a sparse array is an inefficient and inelegant algorithm in solidity.
```
Comment (xklob) [inline contracts/sentinel/guards/FuseWithdrawalGuard.sol:95]: Good explanation, appreciate

**Review by xklob — CHANGES_REQUESTED** [review]
See couple of comments. Overall looks good but would prefer if this was switched to having immutable state (which would also allow you to remove the dependency on CoreRef)

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol (hunk 6 above)
Comment (xklob) [reply contracts/sentinel/guards/FuseWithdrawalGuard.sol:25]: Nvm this is fine.

**Comment by xklob** [comment]
See couple of comments. Overall looks good but would prefer if this was switched to having immutable state (which would also allow you to remove the dependency on CoreRef)

**Comment by eswak** [comment]
lgtm, would like to see an operational deployment next (in the follow-up PR!)

File: contracts/sentinel/guards/FuseWithdrawalGuard.sol (hunk 7 above)
Comment (Joeysantoro) [reply contracts/sentinel/guards/FuseWithdrawalGuard.sol:30]: PCV deposits are always denominated in underlying, I think the naming is fine but can update comment

**Review by xklob — DISMISSED** [review]

File: protocol-configuration/proposalsConfig.ts [hunk 10]
```diff
@@ -1,10 +1,12 @@
 import { ProposalCategory, TemplatedProposalDescription, TemplatedProposalsConfigMap } from '@custom-types/types';
 
+import fip_x from '@proposals/description/fip_x';
+
 const proposals: TemplatedProposalsConfigMap = {
   fuse_withdrawal_guard: {
     deploy: true, // deploy flag for whether to run deploy action during e2e tests or use mainnet state
     totalValue: 0, // amount of ETH to send to DAO execution
-    proposal: tip_117, // full proposal file, imported from '@proposals/description/fip_xx.ts'
+    proposal: fip_x as unknown as TemplatedProposalDescription, // temporary hacky workaround, this field isn't needed but can't be undefined
```
Comment (xklob) [inline protocol-configuration/proposalsConfig.ts]: Let's change the base type of this `proposal` var within the `TemplatedProposalsConfigMap` to include undefined as an acceptable type

**Review by xklob — APPROVED** [review]

**Pull request merged**